Allow servers to be instatiated directly from net.Listener

connor4312 opened the pull request:
What the title says 😛  This allows servers to be instantiated from existing sockets, such that those that are passed over as fds when the process starts.

connor4312 commented:
This actually needs some further work to fit our purposes. I'll close this and reopen it when that's done.

connor4312 commented:
Okay, we're good now. I added functionality to release the listening socket, which closes the Accept() loop and returns the listener. In an RTMP server this method may be called to surrender the listening socket to pass it to a new instance of the server upon restart.

ttaylorr commented on server/server.go:
```diff
@@ -15,33 +16,57 @@ import (
 // Otherwise, a server is returned.
 func New(bind string) (*Server, error) {
-	socket, err := net.Listen("tcp", bind)
+	network := "tcp"
```
Promote to package-level `const`.

ttaylorr commented on server/server.go:
```diff
@@ -62,6 +87,15 @@ func (s *Server) Errs() <-chan error {
+// will no longer be in use.
+func (s *Server) Release() *net.TCPListener {
+	s.release <- struct{}{}
+	<-s.release
```
What does this read do?

ttaylorr commented on server/server.go:
```diff
@@ -73,13 +107,25 @@ func (s *Server) Errs() <-chan error {
+
 	for {
-		conn, err := s.socket.Accept()
+		l.SetDeadline(time.Now().Add(s.deadline))
```
I'm not sure I understand why we have to maintain our own constant deadline that is simply added to the loop each time. Calling `Release()` should call `s.socket.SetDeadline(0)`, which will cause an immediate timeout, even if we're between calls in this loop.

The `s.release` channel is a good idea to keep, that way we only return when the timeout has passed and we wanted it to.

connor4312 replied:
A zero time disables the deadline (https://golang.org/pkg/net/#TCPListener.SetDeadline). I'm not sure deadlines can be modified mid-accept but I'll look at that.

ttaylorr replied:
Wow! I totally thought that I replied to this comment. Anyway, I think the idea is to set the deadline to some instant before `time.Now()` which would cause `Accept()` to return a `net.Error` which you can call `Timeout()` on.

connor4312 commented:
Okay, I fixed mentioned issues, and also happened upon an issue which was causing issues locally and on Travis. When the connection is closed, the Accept loop is never actually closed and continues to spam errors down the Errs() channel (talking about use of a closed network connection). To fix this I moved the server to a state-driven system which, while it is certainly less pretty than a couple of channels, is quite solid and handles most edge cases. The only case it may not handle is Close() and Release() both being called before either of them returns, though that could be avoided by requiring use of Server methods to be synchronous.
